# Tensor 생성
- 파이토치에서 데이터를 저장하는 자료구조
- ndarray와 성격, 사용법이 유사하다.

> **pytorch의 tensor는 숫자 데이터 타입만 지원한다.**


##  원하는 형태(shape) 텐서 생성
- **torch.tensor(자료구조 \[, dtype\])**
    - 지정한 dtype(Data type)에 맞는 Tensor객체를 생성해서 반환한다.
      
## 특정 타입의 Tensor를 직접 생성
- torch.tensor()로 생성하면서 dtype을 지정하면 아래 타입의 Tensor객체가 생성된다.
- 원하는 Type의 Tensor클래스를 이용해 직접 생성해도 된다.
- **torch.FloatTensor(자료구조)**
    - float32 타입 텐서 생성
- **torch.LongTensor(자료구조)**
    - int64 타입 텐서생성
- 그외
    - BoolTensor(bool), CharTensor(int8), ShortTensor(int16), IntTensor(int32), DoubleTensor(float64)
    
## tensor 상태 조회
- **tensor.shape, tensor.size(\[축번호\])**
    -  tensor의 shape조회
- **tensor.dtype, tensor.type()**
    - tensor 원소들의 데이터타입 조회
    - dtype은 **data type**을 type()은 tensor **객체의 클래스 타입**을 반환한다.
- **tensor.ndim, tensor.dim()**  : tensor 차원
- **tensor.numel()**: 전체 원소 개수(numelement)



In [2]:
import torch
import numpy as np

In [3]:
np.array([1, 2, 3], dtype=np.int8)#"int8")

array([1, 2, 3], dtype=int8)

In [4]:
print('torch Data types')   #64bit는 잘 안씀 #np랑 다르게 스트링으로는 못줌.
print("float", torch.float16, torch.float32, torch.float64, torch.float, torch.double)  #torch.float: float32
print("int:", torch.int8, torch.int16, torch.int32, torch.int64, torch.short, torch.int, torch.long)# 가능
print("uint:", torch.uint8, torch.uint16, torch.uint32, torch.uint64)
print("bool:", torch.bool)

torch Data types
float torch.float16 torch.float32 torch.float64 torch.float32 torch.float64
int: torch.int8 torch.int16 torch.int32 torch.int64 torch.int16 torch.int32 torch.int64
uint: torch.uint8 torch.uint16 torch.uint32 torch.uint64
bool: torch.bool


In [5]:
a = torch.tensor([[1,2],[3,4]]) #, dtype=torch.float32)


print("shape:", a.shape, a.size())
print("0축 크기:", a.shape[0], a.size(0))
print("type:", a.type(), a.dtype)  # type(): Tensor 객체 타입. dtype: data type => 둘은 좀 다르다.
print('차원크기:', a.dim(), a.ndim)
print('원소개수:', a.numel())
print("device:", a.device)  # tensor를 다루는 processor(메모리위치) - cpu, cuda(gpu)
# CPU - RAM & GPU(nvida) - VRAM ->  둘 중에 어느 RAM에 올려둘지를 설정해두어야 함.

shape: torch.Size([2, 2]) torch.Size([2, 2])
0축 크기: 2 2
type: torch.LongTensor torch.int64
차원크기: 2 2
원소개수: 4
device: cpu


In [ ]:
aa = torch.tensor(range(10))
print(aa)
aa.type() 

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])


'torch.LongTensor'

In [7]:
#Float, Double(32, 64bit 실수)/Int, Long(32, 64 bit 정수) type Tensor
b = torch.FloatTensor([1,3,7])  #float32
print(b.dtype)
c = torch.IntTensor([10,20,30])    # int32
print(c.dtype)
d = torch.DoubleTensor([1, 2, 3])  # float64
print(d.dtype)
e = torch.LongTensor([10, 20, 30]) # int64
print(e.dtype, e.type())

torch.float32
torch.int32
torch.float64
torch.int64 torch.LongTensor


## 동일한 값을 원소로 가지는 Tensor 생성
- **torch.zeros(\*size), zeros_like(텐서)**: 0으로 구성된 tensor 생성
- **torch.ones(\*size), ones_like(텐서)**: 1로 구성된 tensor생성
- **torch.full(size, fill_value), full_like(텐서, fill_value)**: 지정한 값으로 구성된 tensor생성
  - size
    - shape을 지정한다.
    - zeros, ones는 가변인자이므로 축별 size를 순서대로 서정
    - full 은 tuple/list로 shape을 지정한다.
    

In [18]:
a = torch.zeros(3,2,3) # 3 x 2 x 3
a
a.dtype
# torch.ones(2,3)  # 2 X 3
# torch.full([3,2], fill_value=100) # 3 x 2


torch.float32

In [19]:
# 특정 배열과 동일한 shape의 tensor 생성.
a = torch.tensor([[1, 2],[3, 4]]) # 2 x 2

print(a.shape)
b = torch.zeros_like(a)  # a와 같은 shape, dtype 으로 생성.
# b = torch.ones_like(a)
# b = torch.full_like(a, 20)
print(b.shape, b.dtype)
# b

torch.Size([2, 2])
torch.Size([2, 2]) torch.int64


## 동일한 간격으로 떨어진 값들로 구성된 배열생성
- **torch.arange(start=0, end, step=1)**
- **torch.linspace(start, end, steps,)** : steps - 원소개수

In [22]:
torch.arange(10)  # end. 0 ~ 10-1 1씩 증가
torch.arange(0, 1, 0.1) # 0 ~ 1-0.1, 0.1
torch.arange(10, 1, -1) # 10 ~ 1-(-1), -1

tensor([10,  9,  8,  7,  6,  5,  4,  3,  2])

In [25]:
torch.linspace(0, 10, 5) # 0 ~ 10, 5개원소    #그 안에 5개를 동일한 간격으로
torch.linspace(0, 1, 11)


tensor([0.0000, 0.1000, 0.2000, 0.3000, 0.4000, 0.5000, 0.6000, 0.7000, 0.8000,
        0.9000, 1.0000])

## 빈 tensor 생성
- **torch.empty(\*size)**

In [26]:
torch.empty(3,2,7) # 3 X 2 X 7   # 값이 없을 수 없으니 0에 가까운 아무 값으로 채워넣음.

tensor([[[ 1.0185e-19,  0.0000e+00, -1.0207e+03,  4.5734e-41,  8.9683e-44,
           0.0000e+00,  1.1210e-43],
         [ 0.0000e+00,  1.0189e-19,  0.0000e+00,  1.0442e-08,  4.2728e-05,
           1.3167e-08,  4.2229e-08]],

        [[ 1.4580e-19,  1.1495e+24,  3.0881e+29,  1.5766e-19,  1.8889e+31,
           7.2065e+31,  2.8404e+29],
         [ 2.3089e-12,  1.9421e+31,  2.7491e+20,  6.1949e-04,  1.9421e+31,
           1.5835e-43,  0.0000e+00]],

        [[ 1.0185e-19,  0.0000e+00, -1.0207e+03,  4.5734e-41,  2.6823e+23,
           2.6820e+23,  1.0316e-08],
         [ 4.2696e-08,  5.3603e-08,  4.2658e-08,  8.2638e+20,  2.3053e-12,
           2.6302e+20,  6.1949e-04]]])

## 난수를 이용한 생성

- **torch.rand(\*size)**: 0 ~ 1사이 실수로 구성된 배열을 생성. 각 값은 균등분포를 따른다.
- **torch.randn(\*size)**: 표준정규분포(평균:0, 표준편차:1)를 따르는 실수로 구성된 배열 생성
- **torch.randint(low=0, high, size)**: 지정한 범위의 정수로 구성된 배열 생성
- **torch.randperm(n)**: 0 ~ n-1 사이의 정수를 랜덤하게 섞은 값을 원소로 가지는 배열 생성

In [47]:
torch.manual_seed(0)  # seed 설정
torch.rand(1, 3, 5) # 100 x 3 x 5
torch.randn(30, 3) # 30 X 3  # 표준정규분포(평균0 표준편차1)
torch.randint(1, 100, (3, 3, 6))  # 정수
torch.randperm(100) # 0 ~ 99 를 섞어서 구성   #정수 #index를 만들때 많이 사용 -> 데이터를 섞어서 추출하고 싶을때

tensor([18, 41, 22, 48, 26,  3, 42, 40, 46, 14, 90, 59, 82, 10, 36, 45,  0, 65,
        33, 63, 64, 85, 89, 49, 60, 44, 52, 50, 98,  1, 93, 78, 56, 77,  7, 34,
        67, 29,  5, 86, 20, 91, 15, 97, 73, 68, 31, 80, 55, 92, 53, 37, 58, 47,
        51, 54, 75, 61, 74, 16, 25,  9, 12,  2,  6, 27, 30,  4, 19, 87, 32, 39,
        23, 17, 96, 21, 79, 84, 24, 72, 83, 69, 70, 99, 76, 43, 35,  8, 88, 13,
        62, 95, 28, 94, 38, 11, 71, 81, 66, 57])

In [51]:
a = torch.arange(100)
a
# a 섞기
idx = torch.randperm(100) # 0 ~ 99 섞어서 반환. -> index
idx
b = a[idx]

In [52]:
b

tensor([35, 11, 68, 83, 43, 32, 59, 73, 18, 75, 13, 62, 40, 22, 52, 46, 56, 79,
        94, 51, 85, 48, 92, 71, 16, 10, 31, 81, 97, 90, 29, 19, 95,  7, 27, 78,
        42,  9, 58,  6, 55, 49, 91, 45,  8, 66, 30, 41, 77, 72, 39,  5, 47, 25,
        63, 82, 24, 60, 12, 98, 14, 17, 23, 87, 36, 33, 21, 67, 96, 26, 57, 76,
        15,  1, 88, 44, 69, 84,  3, 89, 61, 64, 80, 28, 53,  0,  4, 99, 34, 86,
        65, 37, 70, 50, 93, 54, 38, 74, 20,  2])

## Tensor gpu/cpu 메모리로 옮기기

- pytorch는 데이터셋인 tensor를 cpu메모리와 gpu 메모리로 서로 옮길 수 있다.
    - 데이터에 대한 연산처리를 어디서 하느냐에 따라 메모리를 선택한다.
    - 장치는 문자열로 설정한다.
        - CPU 사용: "cpu"
        - nvida GPU: "cuda"
        - Apple m1: "mps"
            - pytorch 1.12 부터 지원
- 옮기기
    - tensor 생성시 `device` 파라미터를 이용해 설정
    - `tensor.to(device)`를 이용해 설정
- 현재 실행환경에서 어떤 장비를 사용할 수 있는지 확인
    - nvidia gpu 사용가능확인
        - `torch.cuda.is_available()` - nvida gpu 사용가능 여부
        - `torch.backends.mps.is_available()` - M1 사용가능 여부

In [1]:
import torch
device = 'cuda' if torch.cuda.is_available() else 'cpu'

# device = 'mps' if torch.backends.mps.is_available() else 'cpu'
# device = 'cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu'
device

'cuda'

In [3]:
t = torch.tensor([1, 2, 3], dtype=torch.float32, device=device) #생성할 때 device 지정. 지정 안하면 cpu가 default
print(t, t.device) # tensor가 어느 device에 있는지 확인  # gpu를 여러개로 나눌 수도 있음. cuda:1...2....

tensor([1., 2., 3.], device='cuda:0') cuda:0


In [4]:
t2 = t.to("cpu")  # 다른 device로 옮기기.
print(t2, t2.device)

tensor([1., 2., 3.]) cpu


In [5]:
t3 = t2.to("cuda")
print(t3, t3.device)

tensor([1., 2., 3.], device='cuda:0') cuda:0


## tensor를 상수로 변환
- torch.Tensor객체를 파이썬 상수로 변환.
- tensor객체.item()
    - Scalar(상수) 또는 원소가 하나인 tensor를 python 상수로 변환

In [1]:
import torch

In [2]:
print(torch.tensor(30))   #0차원 텐서

tensor(30)


In [3]:
a = torch.tensor(10)  # 상수(scalar) => 0차원 Tensor
print(a, a.ndim)
print(a.item())  #원소가 하나일 때 item을 쓰면 상수로 뽑아나옴
print(type(a), type(a.item()))

tensor(10) 0
10
<class 'torch.Tensor'> <class 'int'>


In [4]:
b = torch.tensor([[20]]) # 원소가 하나인 N차원 배열
print(b, b.dim())
print(b.item()) #원소가 하나인 배열(텐서) 변환 가능

tensor([[20]]) 2
20


In [6]:
c = torch.tensor([1, 10, 100])
print(c)
# print(c.item()) #원소가 여러개일 경우 Exception발생

tensor([  1,  10, 100])


In [8]:
d = torch.tensor([10], device='cuda')
print(d)
print(d.item())

AssertionError: Torch not compiled with CUDA enabled

## ndarray 호환

- ndarray를 tensor로 생성
    - **torch.tensor(ndarray)**
    - **torch.from_numpy(ndarray)**
- tensor를 ndarray로 변환
    - **tensor.numpy()**
    - tensor가 gpu에 있을 경우 cpu로 옮긴 뒤 변환해야 한다.

In [7]:
import numpy as np
import torch

In [12]:
# ndarray -> tensor
arr = np.arange(1,10)

torch.tensor(arr, dtype=torch.float32)
torch.from_numpy(arr)

tensor([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [13]:
# tensor -> ndarray
t = torch.randn(3,3)
print(t)
t.to("cpu").numpy()   # cpu로 이동한 다음에 numpy 배열로 변경

tensor([[ 1.4469, -0.0119,  0.5355],
        [-1.8660,  0.4020, -1.1341],
        [ 0.5087,  1.1322, -0.4516]])


array([[ 1.4469491 , -0.0118654 ,  0.53551424],
       [-1.8659848 ,  0.402033  , -1.1340578 ],
       [ 0.508675  ,  1.1321652 , -0.45156768]], dtype=float32)

In [ ]:
t2 = torch.randn(2,2, device="cuda")
t2

In [ ]:
t2.numpy()
# t2.to("cpu").numpy()
# t2.cpu().numpy()

# 원소 조회및 변경

## indexing/slicing

- 대부분 Numpy 와 동일
    - **slicing에서 step을 <u>음수로 지정할 수 없다.</u>**


In [16]:
t = torch.randint(-10, 10, (100, ))
t

tensor([  8,   8,   9,   2,   3,  -3,  -3,   8,  -2,   5,   9,   5, -10,  -6,
         -2,  -7,  -9,  -3,  -6,   4,   2,  -3,   2,   7,  -6,   4,   5,   7,
          7, -10,   4,   6,  -9,  -8,  -9,  -8,   6,   8,   3,  -1,   4,  -4,
         -5,  -9,   5,   5,  -1,   3,  -1,  -9,  -1,   1,   4,   0,   8,   7,
          2,   7,   7,   6,  -1,   2,   7,   0,   4,  -6,  -1,  -6,   1,  -6,
         -8,   0,  -7,  -2,   8,   9,  -1,   4,   6,   0,  -6,  -7, -10,   3,
          6,  -8, -10,  -5, -10,  -3,  -5,   1,   6,   0,   3,   7,  -8,   7,
         -8,  -6])

In [18]:
t[0]
t[[1, 5, -1]] # 여러개 조회-> fancy indexing

tensor([ 8, -3, -6])

In [22]:
t[:5]
t[10:15]
t[90:]
t[3:30:3]

####### step 음수 안된다. 그래서 reverse(역순 조회)가 안됨. reverse하려면 flip() 사용
# t[10:1:-2]    # 이런식으로 음수로 표현할 수 없음
t[1:10:2].flip(dims=(0,)) #reverse   # dim=0, 어느축을 기준으로 바꿀건지 설정

tensor([ 5,  8, -3,  2,  8])

In [23]:
t[1:10:2]

tensor([ 8,  2, -3,  8,  5])

## boolean index
- 논리 연산은 &, |, ~ 을 사용한다.

In [24]:
# boolean index
t[t > 0]

tensor([8, 8, 9, 2, 3, 8, 5, 9, 5, 4, 2, 2, 7, 4, 5, 7, 7, 4, 6, 6, 8, 3, 4, 5,
        5, 3, 1, 4, 8, 7, 2, 7, 7, 6, 2, 7, 4, 1, 8, 9, 4, 6, 3, 6, 1, 6, 3, 7,
        7])

In [25]:
t[(t > 0) & (t < 5)]

tensor([2, 3, 4, 2, 2, 4, 4, 3, 4, 3, 1, 4, 2, 2, 4, 1, 4, 3, 1, 3])

In [26]:
# boolean index 처리 함수.
t.masked_select(t > 0)

tensor([8, 8, 9, 2, 3, 8, 5, 9, 5, 4, 2, 2, 7, 4, 5, 7, 7, 4, 6, 6, 8, 3, 4, 5,
        5, 3, 1, 4, 8, 7, 2, 7, 7, 6, 2, 7, 4, 1, 8, 9, 4, 6, 3, 6, 1, 6, 3, 7,
        7])

In [27]:
t

tensor([  8,   8,   9,   2,   3,  -3,  -3,   8,  -2,   5,   9,   5, -10,  -6,
         -2,  -7,  -9,  -3,  -6,   4,   2,  -3,   2,   7,  -6,   4,   5,   7,
          7, -10,   4,   6,  -9,  -8,  -9,  -8,   6,   8,   3,  -1,   4,  -4,
         -5,  -9,   5,   5,  -1,   3,  -1,  -9,  -1,   1,   4,   0,   8,   7,
          2,   7,   7,   6,  -1,   2,   7,   0,   4,  -6,  -1,  -6,   1,  -6,
         -8,   0,  -7,  -2,   8,   9,  -1,   4,   6,   0,  -6,  -7, -10,   3,
          6,  -8, -10,  -5, -10,  -3,  -5,   1,   6,   0,   3,   7,  -8,   7,
         -8,  -6])

In [28]:
# 변경
t[0] = 100
t

tensor([100,   8,   9,   2,   3,  -3,  -3,   8,  -2,   5,   9,   5, -10,  -6,
         -2,  -7,  -9,  -3,  -6,   4,   2,  -3,   2,   7,  -6,   4,   5,   7,
          7, -10,   4,   6,  -9,  -8,  -9,  -8,   6,   8,   3,  -1,   4,  -4,
         -5,  -9,   5,   5,  -1,   3,  -1,  -9,  -1,   1,   4,   0,   8,   7,
          2,   7,   7,   6,  -1,   2,   7,   0,   4,  -6,  -1,  -6,   1,  -6,
         -8,   0,  -7,  -2,   8,   9,  -1,   4,   6,   0,  -6,  -7, -10,   3,
          6,  -8, -10,  -5, -10,  -3,  -5,   1,   6,   0,   3,   7,  -8,   7,
         -8,  -6])

In [29]:
t = torch.arange(1, 10).reshape(3,3)
t

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])

In [32]:
# t[ 0축, 1축 ]
t[1, 2]#.item()   # item()을 안쓰면 tensor 안에다가 넣어줌
t[0, 2].item()

3

In [35]:
t[[1,0], [2,2]]  #(1,2), (0,2)   # 이건 item() 조회 안됨!
# t[[첫번째값, 두번째값]-0축, [첫번째값, 두번째값]-1축]

tensor([6, 3])

# Reshape

## shape 변경
- tensor객체.reshape(\*shape) / view(\*shape) 이용
    - 변환 후 값을 변경하면 원본 배열의 값도 같이 바뀐다.
 > tensor.clone(): tensor를 복제한다.

In [36]:
a=torch.rand(12)
a2 = a.reshape(3,4)
a3 = a.reshape((3,2,2))
a4 = a.reshape((3,2,-1))  #한 개 axis는 -1로 설정가능하고 그럼 계산해서 알아서 설정해 준다.
print(a.shape, a2.size(), a3.shape, a4.shape)


torch.Size([12]) torch.Size([3, 4]) torch.Size([3, 2, 2]) torch.Size([3, 2, 2])


In [ ]:
a5 = a.view(3,4)
a6 = a.view((3,2,2))
a7 = a.view((3,2,-1))  #한개 axis는 -1로 설정가능
print(a.shape, a5.size(), a6.shape, a7.shape)

In [37]:
# a5[0, 0] = 12.1
a2 = a.reshape(3,4)  # a2는 a를 reshape한 형태
a2[0, 1] = 15.1   # a2의 원소를 바꾸면 a도 같이 원소의 값이 변경됨.

print(a2)

tensor([[ 0.6239, 15.1000,  0.5696,  0.5403],
        [ 0.3232,  0.5793,  0.0590,  0.6368],
        [ 0.8061,  0.2889,  0.7944,  0.5383]])


In [42]:
a

tensor([ 0.6239, 15.1000,  0.5696,  0.5403,  0.3232,  0.5793,  0.0590,  0.6368,
         0.8061,  0.2889,  0.7944,  0.5383])

In [43]:
# tensor복사: clone() 메소드
r = a.reshape(3, 4).clone()   #원소를 안바뀌게 하고 싶으면  clone한 뒤에 새로운 변수에 넣고 변경
r
r[0,0] = 100.1
print(a)

tensor([ 0.6239, 15.1000,  0.5696,  0.5403,  0.3232,  0.5793,  0.0590,  0.6368,
         0.8061,  0.2889,  0.7944,  0.5383])


In [41]:
r

tensor([[1.0010e+02, 1.5100e+01, 5.6960e-01, 5.4033e-01],
        [3.2318e-01, 5.7927e-01, 5.9028e-02, 6.3679e-01],
        [8.0614e-01, 2.8891e-01, 7.9435e-01, 5.3827e-01]])

## dummy 축 늘리기

- None을 이용 (numpy의 newaxis 대신 None을 사용한다.)
- unsqueeze(dim=축번호)

In [55]:
import torch
a = torch.tensor([[10,20],[10,20]])
print(a.shape)

a1, a2 = a[None, :], a.unsqueeze(dim=0) # dim=정수만가능
print(a1.shape, a2.shape)

a3, a4 = a[:,:, None], a.unsqueeze(dim=-1)  # 축별로 ':'을 넣어줘야 함.
# a[:, None] # 0축 :, 1축 None, 2번축 지정안함 -> 원래대로 그래서 (2, 1, 2) 가됨.
print(a3.shape, a4.shape)

a5, a6 = a3[:,None,:,:, None],  a3.unsqueeze(dim=1)  # unsqueeze(): dim=정수만가능 (한번에 dummy축은 하나만 추가가능.)
print(a5.shape, a6.shape)

torch.Size([2, 2])
torch.Size([1, 2, 2]) torch.Size([1, 2, 2])
torch.Size([2, 2, 1]) torch.Size([2, 2, 1])
torch.Size([2, 1, 2, 1, 1]) torch.Size([2, 1, 2, 1])


## dummy 축 제거
- squeeze(\[dim=축번호\]) 이용

In [60]:
t = torch.rand(3, 1, 4, 1, 5, 1)
print(t.shape)

r1 = t.squeeze()  #축을 명시 하지 않으면 모두 제거
print(r1.shape)

r2 = t.squeeze(dim=1) # 특정 axis 제거
print(r2.shape)

r3 = t.squeeze(dim=[1,3]) # 여러 axis의 dummy 축 제거
print(r3.shape)

torch.Size([3, 1, 4, 1, 5, 1])
torch.Size([3, 4, 5])
torch.Size([3, 4, 1, 5, 1])
torch.Size([3, 4, 5, 1])


# tensor 합치기
torch.cat([tensorA, tensorB, ...], dim=0) : 기준축 default 는 0
- 합치는 기준 축 이외ㅇ 축의 size가 동일해야 합칠 수 있다.

In [62]:
a = torch.arange(10).reshape(2,5)   # 내가 합치려는 축을 제외한 나머지 축의 수가 같아야함. ex. (3,4,5,6)  (6,4,5,6)  0번 기준으로 합치기 가능
b = torch.arange(10,20).reshape(2,5)
c = torch.arange(20,30).reshape(2,5)
d = torch.arange(10,19).reshape(3,3)

In [63]:
torch.cat([a, b], dim=0)   #0번 축 기준으로 합치니까 나머지 축의 shape가 같아야함.    # 4x5

tensor([[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19]])

In [64]:
torch.cat([a, b, c], dim=0)        # 6X5

tensor([[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]])

In [65]:
torch.cat([a, b], axis=1) # dim 대신 axis사용가능    # 2X10

tensor([[ 0,  1,  2,  3,  4, 10, 11, 12, 13, 14],
        [ 5,  6,  7,  8,  9, 15, 16, 17, 18, 19]])

In [66]:
torch.cat([a, b, c], axis=1)    # 2X15

tensor([[ 0,  1,  2,  3,  4, 10, 11, 12, 13, 14, 20, 21, 22, 23, 24],
        [ 5,  6,  7,  8,  9, 15, 16, 17, 18, 19, 25, 26, 27, 28, 29]])

In [67]:
torch.cat([a, b], axis=-1)  # -1: 마지막 axis  # 2X10

tensor([[ 0,  1,  2,  3,  4, 10, 11, 12, 13, 14],
        [ 5,  6,  7,  8,  9, 15, 16, 17, 18, 19]])

In [69]:
a.shape, d.shape   # 못 합침!!
# 합치는 기준축 이외의 축 size는 같아야 한다.
# torch.cat([a, d])  #dim=0, Error  1축 size가 달라서 Error

(torch.Size([2, 5]), torch.Size([3, 3]))

# 값의 위치(index) 변경
- tensor 원소의 index의 위치를 바꾼다.
- `tensor.transpose(axis1, axis2)`
    - 두 축의 자리만 변경 할 수 있다.
- `tensor.permute(axis1, axis2, axis3, ..)`
    - 두 개 이상의 축 자리를 변경한다.

In [70]:
X = torch.arange(24).reshape(2, 3, 4)
print(X.shape)

y = X.transpose(1, 2)   # 축을 설정해주는 것
print(y.shape)

z = X.permute(2, 0, 1) # n개의 축일 경우에 옮길 대상의 축을 그 위치에 적어주면 됨,
print(z.shape)

torch.Size([2, 3, 4])
torch.Size([2, 4, 3])
torch.Size([4, 2, 3])


In [74]:
z

tensor([[[ 0,  4,  8],
         [12, 16, 20]],

        [[ 1,  5,  9],
         [13, 17, 21]],

        [[ 2,  6, 10],
         [14, 18, 22]],

        [[ 3,  7, 11],
         [15, 19, 23]]])

In [73]:
X

tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

# tensor 연산 및 주요 함수

## element-wise 연산
- tensor와 상수 연산시, tensor와 tensor간 연산시 원소별로 처리한다.
- 행렬곱 연산을 제외하고 tensor간 연산시 피연산자 tensor간에 shape이 같아야 한다.
    - shape이 다를 경우 조건이 맞으면 broadcasting을 한 뒤에 연산한다. (size가 다른 축의 경우 한개의 피연산자 size가 1일 경우 복사하여 shape을 맞춘다.)
    

In [75]:
import torch
a = torch.arange(10).reshape(2,5)
b = torch.arange(10,20).reshape(2,5)
c = torch.arange(50, 55)

print(a)
print(b)
print(c)

tensor([[0, 1, 2, 3, 4],
        [5, 6, 7, 8, 9]])
tensor([[10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19]])
tensor([50, 51, 52, 53, 54])


In [76]:
print(a + 100)
print(a - 100)
print(a < 5)

tensor([[100, 101, 102, 103, 104],
        [105, 106, 107, 108, 109]])
tensor([[-100,  -99,  -98,  -97,  -96],
        [ -95,  -94,  -93,  -92,  -91]])
tensor([[ True,  True,  True,  True,  True],
        [False, False, False, False, False]])


In [77]:
print(a + b)
print(a == b)

tensor([[10, 12, 14, 16, 18],
        [20, 22, 24, 26, 28]])
tensor([[False, False, False, False, False],
        [False, False, False, False, False]])


In [78]:
a.size(), c.size()

(torch.Size([2, 5]), torch.Size([5]))

In [79]:
# broadcasting  # 사이즈가 달라도 broadcasting이 일어남. c -> [[50, 51, 52, 53, 54], [50, 51, 52, 53, 54]] 를 한다음에 a의 각 행과 더함.
print(a + c)

tensor([[50, 52, 54, 56, 58],
        [55, 57, 59, 61, 63]])


## 주요 연산함수

### 주요 상수
#### 자연상수 E

In [81]:
torch.e, np.e, torch.pi, np.pi

(2.718281828459045, 2.718281828459045, 3.141592653589793, 3.141592653589793)

#### torch.nan, torch.inf
- nan: Not a Number, 주로 결측치를 표현한다.
- inf: infinit 무한.
    - torch.inf: 양의 무한
    - -torch.inf: 음의 무한
- torch.isnan(tensor)
    - 원소별 결측치 확인
- torch.isinf(tensor)    
    - 원소별 inf 확인

In [82]:
print(torch.inf > 10000000000000000, torch.inf < 10000000000) #  torch.inf 가 붙으면 어떤 값보다 젤 큼
print(-torch.inf < -1000000000000000000000, -torch.inf > 10) # -torch.inf : 어떤 값보다 젤 작음.


True False
True False


In [83]:
print(torch.log(torch.tensor(-1))) # nan (계산결과가 없으므로-없는값-nan 반환)  # 밑이 e 자연상수는 양수라서 log-1은 없어서 nan이 나옴
print(torch.isnan(torch.tensor([1,2,torch.nan,3,4])))  # nan 여부 확인
print(torch.isinf(torch.tensor([1,2,3,4,torch.inf])))  # inf 여부 확인

tensor(nan)
tensor([False, False,  True, False, False])
tensor([False, False, False, False,  True])


## 주요 연산 함수

In [84]:
x=torch.arange(-4, 5).reshape(3,3)   #  element wise연산
print(x)
print(torch.abs(x)) # 절대값
print(torch.sqrt(torch.abs(x))) #  제곱근
print(torch.exp(x))  # torch.e**x
print(torch.log(torch.abs(x)))
print(torch.log(torch.exp(torch.tensor(1))))  # torch.log() 밑이 e인 로그계산
print(torch.log10(torch.tensor(10)))         # torch.log10() 밑이 10인 로그계산
print(torch.log2(torch.tensor(2)))           # torch.log2() 밑이 2인 로그계산

tensor([[-4, -3, -2],
        [-1,  0,  1],
        [ 2,  3,  4]])
tensor([[4, 3, 2],
        [1, 0, 1],
        [2, 3, 4]])
tensor([[2.0000, 1.7321, 1.4142],
        [1.0000, 0.0000, 1.0000],
        [1.4142, 1.7321, 2.0000]])
tensor([[1.8316e-02, 4.9787e-02, 1.3534e-01],
        [3.6788e-01, 1.0000e+00, 2.7183e+00],
        [7.3891e+00, 2.0086e+01, 5.4598e+01]])
tensor([[1.3863, 1.0986, 0.6931],
        [0.0000,   -inf, 0.0000],
        [0.6931, 1.0986, 1.3863]])
tensor(1.)
tensor(1.)
tensor(1.)


In [85]:
y = x + torch.randn((3,3))
print(y)
print(torch.round(y)) # 반올림
print(torch.round(y, decimals=2)) # 소수점 둘째자리 이하에서 반올림
print(torch.floor(y)) # 내림
print(torch.ceil(y)) # 올림

tensor([[-3.7580, -2.5212, -1.9373],
        [ 0.8168, -0.5812, -0.1111],
        [ 2.9445,  1.3281,  5.3651]])
tensor([[-4., -3., -2.],
        [ 1., -1., -0.],
        [ 3.,  1.,  5.]])
tensor([[-3.7600, -2.5200, -1.9400],
        [ 0.8200, -0.5800, -0.1100],
        [ 2.9400,  1.3300,  5.3700]])
tensor([[-4., -3., -2.],
        [ 0., -1., -1.],
        [ 2.,  1.,  5.]])
tensor([[-3., -2., -1.],
        [ 1., -0., -0.],
        [ 3.,  2.,  6.]])


### 행렬곱
- `@` 연산자 또는 `torch.matmul(tensor1, tensor2)` 함수 이용

In [ ]:
import torch

In [86]:
x = torch.FloatTensor([[1, 2],[3, 4],[5, 6]])

y = torch.FloatTensor([[1, 2],[1, 2],])
x.size(), y.shape

(torch.Size([3, 2]), torch.Size([2, 2]))

In [87]:
z1 = x @ y
z2 = torch.matmul(x, y) # matrix multiply
print(z1.shape, z2.shape)
print(z1)
print(z2)

torch.Size([3, 2]) torch.Size([3, 2])
tensor([[ 3.,  6.],
        [ 7., 14.],
        [11., 22.]])
tensor([[ 3.,  6.],
        [ 7., 14.],
        [11., 22.]])


In [88]:
# 3차원이상의 행렬곱을 하고 싶을때:ㅣ Batch 행렬곱(Batch matrix muliplication) - bmm()
# 피연산자로 두개의 3차원 tensor를 axis (1, 2)를 기준으로 행렬곱을 처리한다.
# 4x2 가 3개씩, 2x5가 3개씩 이걸 행렬곱을 하면, 4x5가 3개가 됨.
import torch
x = torch.FloatTensor(3,4,2)
y = torch.FloatTensor(3,2,5)
z = torch.bmm(x, y)
z.shape


torch.Size([3, 4, 5])

## 기술통계함수

In [89]:
X=torch.randn(3,4)
print(X)

tensor([[-0.8977,  1.6361,  0.8110,  0.2115],
        [ 0.0864, -0.9572,  1.5467,  0.7313],
        [ 1.0972,  2.7566,  0.0551,  0.9365]])


In [92]:
print(torch.sum(X))  # default: 전체기준으로 계산. dim=None
print(torch.sum(X, dim=1)) #dim/axis 지정: 지정한 axis의 index가 다른 값끼리 계산.
print(torch.sum(X, dim=1, keepdims=True)) # 기술통계함수들을 실행하면 차원이 줄어든다. keepdims=True로 하면 차원유지

tensor(8.0134)
tensor([1.7608, 1.4073, 4.8454])
tensor([[1.7608],
        [1.4073],
        [4.8454]])


In [93]:
print(torch.mean(X))
print(torch.mean(X, dim=0))
print(torch.mean(X, dim=0, keepdims=True))

tensor(0.6678)
tensor([0.0953, 1.1452, 0.8043, 0.6264])
tensor([[0.0953, 1.1452, 0.8043, 0.6264]])


In [94]:
print(torch.std(X)) # standard deviation 표준 편차
print(torch.var(X)) # variance
print(torch.var(X, dim=0))

tensor(1.0603)
tensor(1.1243)
tensor([0.9949, 3.6288, 0.5562, 0.1396])


In [ ]:
# tensor.메소드()
print(X.sum(dim=1, keepdims=True))
print(X.mean(dim=1, keepdims=True))
print(X.std())

In [ ]:
X

In [100]:
print(torch.max(X)) # 전체 기준 max값 계산.
print(torch.max(X, dim=0))  # 축을 지정하면 return_types.max 타입객체로 반환. values = :max값과 indices = : max값의 index를 묶어서 반환
print(torch.max(X, dim=1))
print(torch.max(X, dim=1).values, torch.max(X, dim=1).indices, sep=" || ")    # values값이랑 indices 값을 따로 뽑을 수 있음


print(torch.max(X, dim=0, keepdims=True))  #keepdims=True : 차원(rank)를 유지
print(torch.max(X, dim=1, keepdims=True))

tensor(2.7566)
torch.return_types.max(
values=tensor([1.0972, 2.7566, 1.5467, 0.9365]),
indices=tensor([2, 2, 1, 2]))
torch.return_types.max(
values=tensor([1.6361, 1.5467, 2.7566]),
indices=tensor([1, 2, 1]))
tensor([1.6361, 1.5467, 2.7566]) || tensor([1, 2, 1])
torch.return_types.max(
values=tensor([[1.0972, 2.7566, 1.5467, 0.9365]]),
indices=tensor([[2, 2, 1, 2]]))
torch.return_types.max(
values=tensor([[1.6361],
        [1.5467],
        [2.7566]]),
indices=tensor([[1],
        [2],
        [1]]))


In [102]:
print(torch.min(X))
print(torch.min(X, dim=0)) # return_types.min 타입객체로 반환. min값과 min값의 index를 묶어서 반환
print(torch.min(X, dim=1))

tensor(-0.9572)
torch.return_types.min(
values=tensor([-0.8977, -0.9572,  0.0551,  0.2115]),
indices=tensor([0, 1, 2, 0]))
torch.return_types.min(
values=tensor([-0.8977, -0.9572,  0.0551]),
indices=tensor([0, 1, 2]))


In [104]:
print(torch.min(X, dim=1).values)
print(torch.min(X, dim=1).indices)

tensor([-0.8977, -0.9572,  0.0551])
tensor([0, 1, 2])


In [110]:
print(torch.argmax(X))      # 전체 중 가장 큰 값의 index -> flatten 한 뒤에 index를 변환
print(torch.argmax(X, dim=0)) # 각 열에서 가장 큰 애가 존재하는 인덱스
print(torch.argmax(X, dim=1)) # 각 행에서 가장 큰 애가 존재하는 인덱스

tensor(9)
tensor([2, 2, 1, 2])
tensor([1, 2, 1])


In [111]:
a = torch.max(X, dim=0)
print(a.values)
print(a.indices)

tensor([1.0972, 2.7566, 1.5467, 0.9365])
tensor([2, 2, 1, 2])


# autograd(자동미분)
- 자동 미분을 이용해 gradient(미분계수)를 계산하는 pytorch system.
- 딥러닝 모델에서 weight와 bias tensor들(Parameter)은 backpropagation(역전파)를 이용해 gradient를 구해서 loss가 줄어드는 방향으로 update를 하게된다.
- pytorch는 이런 미분 수행을 자동으로 처리해 준다.
    - gradient(기울기)를 구한다는 것은 미분을 한다는 것을 말한다.
- tensor가 미분 가능하려면(gradient 계산 대상 변수) `requires_grad=True` 로 설정되어 있어야 한다. (default: False) [필수]


- [추론] forward : 도함수 정의/ [loss] backward: gradient 계산

> - **미분**
>   -  (순간) 변화율을 계산한다.
>   - $$\frac{\partial y}{\partial x}$$
>       - x에 대한 y의 변화율. 즉 x가 변하면 y는 얼마나 변할까?

In [112]:
import torch

In [183]:
x = torch.tensor([1.0], requires_grad=True) # Gradiens를 계산할 대상인지 여부
print(x.requires_grad)  #default = false
## True: x를 이용한 계산을 할때 gradient함수(도함수)를 정의해서 저장
# x.requires_grad = True # 따로 변경 가능

True


In [182]:
# Tensor의 속성
## data - Tensor가 가지는 값들을 저장하느 변수(속성)
print(x.data) # tensor([1.])
## grad - Gradient 값을 저장하는 변수(속성)
print(x.grad)

tensor([1.])
tensor([3.])


In [189]:
y = x ** 2
y # tensor([1.], grad_fn=<PowBackward0>)   # pow ; 제곱의 도함수

tensor([1.], grad_fn=<PowBackward0>)

In [185]:
y.grad_fn    # gradient계산함수(도함수)가 저장된 변수(속성)  # dy / dx (x(requires_grad = True)에 대한 y의 변화율)

In [190]:
#gradient 계산
y.backward() # dy/dx 계산 -> 그 결과를 x.grad에 저장

In [ ]:
x.data, x.grad, y.grad_fn

# X -> data = [1.0], require_grad = True, grad_fn = None
# y = x**2 -> data= [1.0], grad_fn = 2x (X에서 requrie_grad=True로 설정했으니깐)
# y.backward() 도함수를 이용해서 계산

(tensor([1.]), tensor([2.]), <PowBackward0 at 0x78f96a0591e0>)

In [191]:
a = torch.tensor([2.0], requires_grad=True)
b = torch.tensor([3.2], requires_grad=False)   # r_grad = False

y = a - b+ 3 #grad_fn 정의
y.backward() # grad를 계산- a에 대해서만 계산.

In [133]:
y.grad_fn, a.data, a.grad

(<AddBackward0 at 0x78f969a40370>, tensor([2.]), tensor([1.]))

In [135]:
b.grad  # 아무것도 없음

In [169]:
# 간단한 경사하강법 구현
input_data = torch.tensor([30.2])
output_data = torch.tensor([10.5])
weight = torch.tensor([0.2], requires_grad=True)
y_pred = weight * input_data
loss = (output_data - y_pred)**2  # 오차계산-MSE 같은 SE   #이것도 grad가 생긴 이유는 y_pred안에 있는 weight가 requires_grad = True 이기 때문

In [145]:
y_pred, loss, y_pred.grad_fn, loss.grad_fn

(tensor([6.0400], grad_fn=<MulBackward0>),
 tensor([19.8916], grad_fn=<PowBackward0>),
 <PowBackward0 at 0x78f969dbf130>)

In [167]:
y_pred.backward()

In [170]:
loss.backward()  #loss를 백워드하면, y_pred의 도함수를 먼저 계산한 다음에 loss의 도함수를 계산해서 합성함수(즉, 두 기울기를 곱한 값)로 계산
# loss = (output_data - (weight*input_data))**2
# weight.grad

In [171]:
weight.grad

tensor([-269.3840])

In [164]:
new_weight = weight.data - weight.grad * 0.0001  #이걸로 weight의 값을 계속해서 줄여줌. gradient가 0이 될때까지.
new_weight

tensor([0.2269])

## torch.no_grad()
- no_grad()를 with 구문에서 연산을 할 경우 with block 내에서 requires_grad=True로 설정된 변수를 이용한 계산식이 있더라도 grad_fn을 생성하지 않는다. 즉 gradient를 계산하지 않는다.
- 딥러닝 모델을 평가 하거나 추론을할 때는 gradient를 계산할 필요가 없기 때문에 no_grad 구문을 사용한다.


In [175]:
x = torch.tensor([1.0], requires_grad = True)
y = x**2
print("결과:", y.data)
print("grad_fn", y.grad_fn)

결과: tensor([1.])
grad_fn <PowBackward0 object at 0x78f969e7b730>


In [176]:
with torch.no_grad():     #추론하거나 평가할때는 gradient가 필요없기 때문에 no_grad 같은 with block 안에서 진행하면 됨.
    y = x**2
    print("결과:", y.data)
    print("grad_fn", y.grad_fn)

결과: tensor([1.])
grad_fn None


## gradient 값 초기화
- backward() 에서 계산되어 저장된 gradient값을 초기화한다.
- 반복 학습을 할 경우 gradient가 누적되는 문제가 발생한다. 그래서 한번 학습이 끝나면 다음 학습 전에 초기화하는 작업을 해야 한다.

In [177]:
x = torch.tensor([1.0], requires_grad= True)
print(x.grad)
y = x**2
y.backward()
print(x.grad)

None
tensor([2.])


In [180]:
# x를 이용한 새로운 계산
x.grad = None  # 초기화
z = x ** 3
z.backward()

In [181]:
print(x.grad)  # 새로운 계산을 할때 grad 값을 초기화하지 않으면 앞에 계산했던 x.grad와 더해짐. # 2+3 tensor([5.])

tensor([3.])
